File: QentaPayAndroid/app/src/main/java/com/example/qentapayandroid/MainActivity.java
package com.example.qentapayandroid;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.widget.Button;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;

public class MainActivity extends AppCompatActivity {

    private Button payButton;
    public static final String NGROK_URL = "https://xib52gsaxp67.share.zrok.io";

    private final ActivityResultLauncher<Intent> paymentLauncher = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(),
            result -> {
                if (result.getResultCode() == RESULT_OK && result.getData() != null) {
                    Uri data = result.getData().getData();
                    handlePaymentResult(data);
                } else {
                    Toast.makeText(this, "Vorgang abgebrochen", Toast.LENGTH_SHORT).show();
                }
            }
    );

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        payButton = findViewById(R.id.btnGooglePay);

        payButton.setOnClickListener(v -> {
            //switch for Webview vs Chrome Custom Tabs
            boolean isUsingWebView = false;
            Intent intent = null;

            if (isUsingWebView) {
                intent = new Intent(MainActivity.this, PaymentActivityWebView.class);
            }
            else {
                intent = new Intent(MainActivity.this, PaymentActivityChromeCustomTabs.class);
            }
            paymentLauncher.launch(intent);
        });
    }

    private void handlePaymentResult(Uri data) {
        if (data == null)
            return;

        String status = data.getQueryParameter("paymentState");
        if (status == null) {
            Toast.makeText(this, "Status unbekannt", Toast.LENGTH_SHORT).show();
            return;
        }

        switch (status) {
            case "SUCCESS":
                Toast.makeText(this, "Bezahlung erfolgreich!", Toast.LENGTH_LONG).show();
                break;
            case "NOTOK":
                Toast.makeText(this, "Abbruch durch System", Toast.LENGTH_SHORT).show();
                break;
            case "FAILURE":
                Toast.makeText(this, "Fehler aufgetreten.", Toast.LENGTH_LONG).show();
                break;
            default:
                Toast.makeText(this, "Vorgang beendet (" + status + ")", Toast.LENGTH_SHORT).show();
                break;
        }
    }
}

File: QentaPayAndroid/app/src/main/java/com/example/qentapayandroid/MainActivity.kt
package com.example.qentapayandroid

import android.annotation.SuppressLint
import android.net.Uri
import android.os.Bundle
import android.view.View
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.Button
import android.widget.Toast
import androidx.activity.OnBackPressedCallback
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity() {

    private lateinit var webView: WebView
    private lateinit var payButton: Button

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.activity_main)

        payButton = findViewById(R.id.btnGooglePay)
        webView = findViewById(R.id.webView)

        setupWebView()
        setupBackButtonHandler()

        payButton.setOnClickListener {
            startPaymentMockup()
        }
    }

    @SuppressLint("SetJavaScriptEnabled")
    private fun setupWebView() {
        val settings = webView.settings
        settings.javaScriptEnabled = true
        settings.domStorageEnabled = true

        webView.webViewClient = object : WebViewClient() {
            override fun shouldOverrideUrlLoading(view: WebView?, request: WebResourceRequest?): Boolean {
                val url = request?.url.toString()

                if (url.contains("maskedPan")) {
                    handlePaymentResult(Uri.parse(url))
                    return true
                }

                // Solange "allclear" NICHT in der URL steht,
                // laden wir alles ganz normal weiter (auch Hobex-Seiten).
                return false
            }
        }
    }

    private fun startPaymentMockup() {
        payButton.visibility = View.GONE
        webView.visibility = View.VISIBLE

        // Deine Start-URL
        val url = "https://8a47d0cec79c.ngrok-free.app"
        webView.loadUrl(url)
    }

    private fun handlePaymentResult(data: Uri) {
        webView.visibility = View.GONE
        payButton.visibility = View.VISIBLE

        // FIX: Hier laden wir "about:blank".
        // Wenn du hier wieder die Start-URL lädst, startet die Zahlung im Hintergrund neu!
        webView.loadUrl("about:blank")

        val status = data.getQueryParameter("paymentState")

        when (status) {
            "SUCCESS" -> Toast.makeText(this, "Bezahlung erfolgreich!", Toast.LENGTH_LONG).show()
            "NOTOK" -> Toast.makeText(this, "Abbruch", Toast.LENGTH_SHORT).show()
            "FAILURE" -> Toast.makeText(this, "Fehler aufgetreten.", Toast.LENGTH_LONG).show()
            // Falls kein Status-Parameter da ist, aber "allclear" gefunden wurde:
            else -> Toast.makeText(this, "Vorgang beendet (All Clear)", Toast.LENGTH_SHORT).show()
        }
    }

    private fun setupBackButtonHandler() {
        onBackPressedDispatcher.addCallback(this, object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                if (webView.visibility == View.VISIBLE) {
                    if (webView.canGoBack()) {
                        webView.goBack()
                    } else {
                        // Wenn man nicht mehr zurück kann -> Abbruch
                        webView.visibility = View.GONE
                        payButton.visibility = View.VISIBLE
                        webView.loadUrl("about:blank")
                        Toast.makeText(this@MainActivity, "Abgebrochen", Toast.LENGTH_SHORT).show()
                    }
                } else {
                    isEnabled = false
                    onBackPressedDispatcher.onBackPressed()
                }
            }
        })
    }
}